crates/worker/src/partition/mod.rs

```diff
@@ -337,62 +347,72 @@ where
-                        }
+                    for (lsn, envelope) in records_buffer.drain(..) {
+                        let command_start = Instant::now();
 
-                        let is_leader = self.leadership_state.on_announce_leader(announce_leader, &mut partition_storage).await?;
+                        trace!(%lsn, "Processing bifrost record for '{}': {:?}", envelope.command.name(), envelope.header);
 
-                        Span::current().record("is_leader", is_leader);
+                        let leadership_change = self.apply_record(
+                            lsn,
+                            envelope,
+                            &mut transaction,
+                            &mut action_collector).await?;
 
-                        if is_leader {
-                            self.status.effective_mode = RunMode::Leader;
+                        if let Some((header, announce_leader)) = leadership_change {
+                            apply_record_latency.record(command_start.elapsed());
+
+                            let commit_start = Instant::now();
+                            // commit all changes so far, this is important so that the actuators see all changes
+                            // when becoming leader.
+                            transaction.commit().await?;
+                            record_batch_commit_latency.record(commit_start.elapsed());
+
+                            // We can ignore all actions collected so far because as a new leader we have to instruct the
+                            // actuators afresh.
+                            action_collector.clear();
+
+                            self.status.last_observed_leader_epoch = Some(announce_leader.leader_epoch);
+                            if let Source::Processor { node_id, .. } = header.source {
+                                // all new AnnounceLeader messages should come from a PartitionProcessor
+                                self.status.last_observed_leader_node = Some(node_id);
+                            } else if announce_leader.node_id.is_some() {
+                                // older AnnounceLeader messages have the announce_leader.node_id set
+                                self.status.last_observed_leader_node = announce_leader.node_id;
+                            }
+
+                            let is_leader = self.leadership_state.on_announce_leader(announce_leader, &mut partition_storage).await?;
+
+                            Span::current().record("is_leader", is_leader);
+
+                            if is_leader {
+                                self.status.effective_mode = RunMode::Leader;
+                            } else {
+                                self.status.effective_mode = RunMode::Follower;
+                            }
+
+                            transaction = partition_storage.create_transaction();
```

Comment:
do we need to `continue` here?

Reply:
Yes, we will continue here if there are more commands in the batch to process.

Reply:
I'm not sure I understand. What I mean is, do we need to fall-through to line 406 in this condition, or should we call `continue;` after creating the transaction?

Reply:
We need to `continue`. Wouldn't the code be the same with `continue;` or w/o?

Reply:
Correct. For some reason I was blurred by the nesting and indentation.